### PR TITLE
[x86/Linux] Fix Dacp structure size mismatch

### PR DESCRIPTION
**Pull request by seanshpark**
Fix compile error for x86/Linux
- add padding with macro
- Fix "Dacp structs cannot be modified due to backwards compatibility" error

### CODE REVIEW

**Comment by seanshpark** [comment]
CC @parjong

**Comment by seanshpark** [comment]
Fixes #8369
@janvorli , @jkotas , @BruceForstall PTAL

**Comment by BruceForstall** [comment]
The approach seems reasonable to me. But @janvorli or @jkotas or @noahfalk should make the call.

**Comment by janvorli** [comment]
@seanshpark I am not sure I understand the issue. Why would this be different from Windows x86 or ARM on Windows on Linux?

**Comment by janvorli** [comment]
What I mean is that the C/C++ compiler should align struct members to their natural alignment and based on your change, it is not happening.

**Comment by janvorli** [comment]
@seanshpark, I have tried to compile a simple piece of code with a structure to see the layout using clang with -m32 option to generate x86 code:
```C++
#include <stdio.h>

struct A
{
  char a;
  short b;
  int c;
  long long d;
};

void f()
{
	printf("%d %d %d", __builtin_offsetof(A, b), __builtin_offsetof(A, c), __builtin_offsetof(A, d)); 
}
```

And here is the generated code that shows that all members are aligned on their natural boundaries, as I would expect (see the numbers 2, 4, 8 - these are the offset of b, c and d members):
```asm
        push    ebp
        mov     ebp, esp
        push    esi
        sub     esp, 36
        lea     eax, [.L.str]
        mov     ecx, 2
        mov     edx, 4
        mov     esi, 8
        mov     dword ptr [esp], eax
        mov     dword ptr [esp + 4], 2
        mov     dword ptr [esp + 8], 4
        mov     dword ptr [esp + 12], 8
        mov     dword ptr [ebp - 8], esi # 4-byte Spill
        mov     dword ptr [ebp - 12], ecx # 4-byte Spill
        mov     dword ptr [ebp - 16], edx # 4-byte Spill
        call    printf
        mov     dword ptr [ebp - 20], eax # 4-byte Spill
        add     esp, 36
        pop     esi
        pop     ebp
        ret
```
So I believe there is likely some other issue - maybe some of the data types used in those structs are defined incorrectly or something like that.

**Comment by seanshpark** [comment]
@janvorli , I've updated with using `MSLAYOUT` macro. I didn't add to all structures but only having error in size checking. If it is better adding to all please let me know.

**Comment by seanshpark** [comment]
CI did not check with Linux/Unix x64. I think it would be safe to do so.

**Comment by seanshpark** [comment]
@dotnet-bot test All please

**Comment by seanshpark** [comment]
@dotnet-bot test this please

**Comment by seanshpark** [comment]
@janvorli , could you take a look again please?

**Comment by seanshpark** [comment]
Stange, now the other CI's are running :)

File: src/inc/dacprivate.h [hunk 1]
```diff
@@ -20,6 +20,14 @@
 #include <msodw.h>
 #endif // FEATURE_PAL
 
+#if !defined(MSLAYOUT)
```
Comment (janvorli) [inline src/inc/dacprivate.h]: I wonder if there could be a common place to put this macro. I am not thrilled by having the same definition at two places - here and in the src/debug/inc/common.h.

File: src/inc/dacprivate.h [hunk 2]
```diff
@@ -279,7 +287,7 @@ struct DacpMethodTableData : ZeroInit<DacpMethodTableData>
 #define CLRSECURITYHOSTED                           0x80
 #define CLRHOSTED           0x80000000
 
-struct DacpThreadStoreData : ZeroInit<DacpThreadStoreData>
+struct MSLAYOUT MSLAYOUT DacpThreadStoreData : ZeroInit<DacpThreadStoreData>
```
Comment (seanshpark) [inline src/inc/dacprivate.h]: Oops

**Comment by seanshpark** [comment]
> add the MSLAYOUT to all the structures

Yes Sure!

File: src/inc/dacprivate.h (hunk 1 above)
Comment (seanshpark) [inline src/inc/dacprivate.h]: Also agree to place at one place. Do you have any place in mind? or a new file?

Comment (seanshpark) [inline src/inc/dacprivate.h]: How about adding new `src/inc/dacstructure.h` containing comments and MSLAYOUT definition? or `daclayout.h`.

Comment (janvorli) [inline src/inc/dacprivate.h]: I did a little experiment - putting the definition only to dacprivate.h and adding `#include "dacprivate.h"` to src/debug/inc/common.h seems to work.

Comment (janvorli) [inline src/inc/dacprivate.h]: Btw, if you do this, could you please copy the code from the common.h as is, including the comment? Also, you don't need to add extra check for __clang__, the _GNU_C_ is good enough (defined for clang too).

Comment (seanshpark) [inline src/inc/dacprivate.h]: OK, thank you!

**Pull request merged**